### PR TITLE
DFU target smp & MCUmgr SMP client unit tests

### PR DESCRIPTION
**Pull request by jheiskan81**
Added unit test for validate DFU Target SMP API. 
Added unit test for validate mcumgr_smp_client API.
Added image list count to 0 at dfu_target_smp_client_init()

### CODE REVIEW

File: tests/subsys/dfu/dfu_target/smp/CMakeLists.txt
Comment (sigvartmh) [inline tests/subsys/dfu/dfu_target/smp/CMakeLists.txt:32]: is there a reason why these are provided through Cmake and not through the YAML?

Comment (jheiskan81) [reply tests/subsys/dfu/dfu_target/smp/CMakeLists.txt:32]: These need to define here that build is not failing. Enabling CONFIG_TARGET_SMP=y was causing a the other problem. We have done similar for other Unit test also.

Comment (sigvartmh) [reply tests/subsys/dfu/dfu_target/smp/CMakeLists.txt:32]: then if it fails for a specific platform you'll have to just filter the allow list.

**Review by sigvartmh — APPROVED** [review]

File: tests/subsys/dfu/dfu_target/smp/testcase.yaml
Comment (gopiotr) [inline tests/subsys/dfu/dfu_target/smp/testcase.yaml:3]: Please add `integration_platforms` option:
```yaml
integration_platforms:
  - native_posix
```

File: tests/subsys/net/lib/mcumgr_smp_client/testcase.yaml
Comment (gopiotr) [inline tests/subsys/net/lib/mcumgr_smp_client/testcase.yaml:3]: Please add `integration_platforms` option:
```yaml
integration_platforms:
  - native_posix
```

**Review by gopiotr — COMMENTED** [review]
Please add to this PR commit with entry into CODEOWNERS file (https://github.com/nrfconnect/sdk-nrf/blob/main/CODEOWNERS) with path to test `/tests/subsys/net/lib/mcumgr_smp_client/` and person/people responsible for this test.

**Comment by jheiskan81** [comment]
@gopiotr Thanks for review. I have updated PR based on your comments.

**Comment by jheiskan81** [comment]
@gopiotr can you review my changes.

**Comment by gopiotr** [comment]
> @gopiotr can you review my changes.

@jheiskan81 - thank you for changes which I asked for. It looks okay for me. I was asked for review in this PR due to the gap in CODEOWNERS file and I try to "guard" this file content (at least in a `tests` area). But I have no knowledge about `dfu` and `mcumgr` and I can't say if tests which you made are okay.

So, I will not block this PR, but it will be good if someone else with more competence could take look at this. 

@de-nordic @hakonfam please review this or pass it on to someone else.

File: tests/subsys/net/lib/mcumgr_smp_client/src/main.c
Comment (gopiotr) [inline tests/subsys/net/lib/mcumgr_smp_client/src/main.c:127]: `Ztest` framework executes tests in alphanumerical order. At first glance it seems to me that this `init` test should be executed at the beginning (but I didn't analyze it deeper, so maybe I'm wrong). When I run those tests locally I can see that test execution order is as below:
```
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_download] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_erase_and_confirm] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_init] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_reset] duration = 0.000 seconds
DEBUG   - OUTPUT:  - PASS - [mcumgr_smp_client_test.test_schedule_image] duration = 0.000 seconds
```
`test_init` test is executed as a third test. Is it intentionally? If not, maybe you should consider changing this `test_init` test as a `ztest_suite_setup_t` function to guarantee that those actions will be executed always at the beginning of test suite.

Comment (jheiskan81) [reply tests/subsys/net/lib/mcumgr_smp_client/src/main.c:127]: I can add init to ZTEST_SUITE() pre action function so Then it will be done properly. Good point. I can change that easily.

**Comment by jheiskan81** [comment]
@gopiotr I move test_init inside ztest_suite_setup_t .

**Review by SeppoTakalo — APPROVED** [review]

**Comment by carlescufi** [comment]
Waiting for @nordicjm to take a look.

File: tests/subsys/dfu/dfu_target/smp/src/os_gr_stub.c
Comment (nordicjm) [inline tests/subsys/dfu/dfu_target/smp/src/os_gr_stub.c:43]: define?

File: tests/subsys/dfu/dfu_target/smp/src/os_gr_stub.c
Comment (nordicjm) [inline tests/subsys/dfu/dfu_target/smp/src/os_gr_stub.c:75]: As above

**Review by nordicjm — COMMENTED** [review]
minor nits.

**Review by nordicjm — APPROVED** [review]

**Comment by jheiskan81** [comment]
rebased

**Pull request merged**
